File: packages/typescript/examples/tipo-cambio-app.ts
#!/usr/bin/env npx ts-node
/**
 * Tipo de Cambio App - TypeScript
 *
 * Consulta y convierte USD/MXN usando los catálogos de Banxico.
 *
 * Uso:
 *   npx ts-node examples/tipo-cambio-app.ts
 */

import { TipoCambioUsdCatalog } from '../src/catalogs/banxico/tipo-cambio-usd';

function printHeader(): void {
  console.log();
  console.log('╔══════════════════════════════════════════════════════╗');
  console.log('║     💵 Tipo de Cambio USD/MXN - catalogmx            ║');
  console.log('╚══════════════════════════════════════════════════════╝');
  console.log();
}

function formatCurrency(amount: number, decimals = 2): string {
  return amount.toLocaleString('es-MX', {
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals,
  });
}

async function showActual(): Promise<void> {
  console.log('  📊 Tipo de Cambio Actual');
  console.log('  ─────────────────────────');

  const actual = TipoCambioUsdCatalog.getActual();

  if (!actual) {
    console.log('  ❌ No se pudo obtener el tipo de cambio');
    return;
  }

  console.log(`  📅 Fecha:        ${actual.fecha}`);
  console.log(`  💱 Tipo cambio:  $${formatCurrency(actual.tipo_cambio, 4)} MXN por 1 USD`);
  console.log();
  console.log('  Ejemplos de conversión:');
  console.log(`    $1 USD     = $${formatCurrency(actual.tipo_cambio)} MXN`);
  console.log(`    $100 USD   = $${formatCurrency(actual.tipo_cambio * 100)} MXN`);
  console.log(`    $1,000 USD = $${formatCurrency(actual.tipo_cambio * 1000)} MXN`);
  console.log();
  console.log(`    $100 MXN   = $${formatCurrency(100 / actual.tipo_cambio)} USD`);
  console.log(`    $1,000 MXN = $${formatCurrency(1000 / actual.tipo_cambio)} USD`);
}

function showHistorico(fecha: string): void {
  console.log(`  📅 Tipo de Cambio del ${fecha}`);
  console.log('  ─────────────────────────────');

  const tc = TipoCambioUsdCatalog.getByFecha(fecha);

  if (!tc) {
    console.log(`  ❌ No hay datos para ${fecha}`);
    console.log('     (Solo días hábiles tienen tipo de cambio)');
    return;
  }

  console.log(`  💱 Tipo cambio: $${formatCurrency(tc.tipo_cambio, 4)} MXN/USD`);
}

function convertir(cantidad: number, direccion: 'usd_mxn' | 'mxn_usd'): void {
  const actual = TipoCambioUsdCatalog.getActual();

  if (!actual) {
    console.log('  ❌ No se pudo obtener el tipo de cambio');
    return;
  }

  if (direccion === 'usd_mxn') {
    const resultado = cantidad * actual.tipo_cambio;
    console.log(`  💵 $${formatCurrency(cantidad)} USD = $${formatCurrency(resultado)} MXN`);
  } else {
    const resultado = cantidad / actual.tipo_cambio;
    console.log(`  🇲🇽 $${formatCurrency(cantidad)} MXN = $${formatCurrency(resultado)} USD`);
  }
  console.log(`     (Tipo de cambio: ${formatCurrency(actual.tipo_cambio, 4)})`);
}

function showEstadisticas(): void {
  console.log('  📊 Estadísticas Recientes');
  console.log('  ─────────────────────────');

  const datos = TipoCambioUsdCatalog.getAll();

  if (datos.length === 0) {
    console.log('  ❌ No hay datos disponibles');
    return;
  }

  // Últimos 30 registros
  const ultimos = datos.slice(-30);
  const valores = ultimos.map(d => d.tipo_cambio);

  const minimo = Math.min(...valores);
  const maximo = Math.max(...valores);
  const promedio = valores.reduce((a, b) => a + b, 0) / valores.length;
  const primero = valores[0];
  const ultimo = valores[valores.length - 1];
  const variacion = ((ultimo - primero) / primero) * 100;

  console.log(`  Registros analizados: ${ultimos.length}`);
  console.log(`  Período: ${ultimos[0].fecha} a ${ultimos[ultimos.length - 1].fecha}`);
  console.log();
  console.log(`  Mínimo:    $${formatCurrency(minimo, 4)}`);
  console.log(`  Máximo:    $${formatCurrency(maximo, 4)}`);
  console.log(`  Promedio:  $${formatCurrency(promedio, 4)}`);
  console.log(`  Variación: ${variacion >= 0 ? '+' : ''}${formatCurrency(variacion)}%`);
}

// Main
async function main(): Promise<void> {
  printHeader();

  const args = process.argv.slice(2);
  const comando = args[0]?.toLowerCase() || 'actual';

  switch (comando) {
    case 'actual':
      await showActual();
      break;

    case 'historico':
      const fecha = args[1] || new Date().toISOString().split('T')[0];
      showHistorico(fecha);
      break;

    case 'usd':
      const cantidadUsd = parseFloat(args[1] || '100');
      convertir(cantidadUsd, 'usd_mxn');
      break;

    case 'mxn':
      const cantidadMxn = parseFloat(args[1] || '1000');
      convertir(cantidadMxn, 'mxn_usd');
      break;

    case 'stats':
      showEstadisticas();
      break;

    default:
      console.log('  Uso:');
      console.log('    npx ts-node tipo-cambio-app.ts actual');
      console.log('    npx ts-node tipo-cambio-app.ts historico 2024-06-15');
      console.log('    npx ts-node tipo-cambio-app.ts usd 100');
      console.log('    npx ts-node tipo-cambio-app.ts mxn 1000');
      console.log('    npx ts-node tipo-cambio-app.ts stats');
  }

  console.log();
}

main().catch(console.error);

File: packages/dart/example/tipo_cambio_app.dart
#!/usr/bin/env dart
/// Tipo de Cambio App - Dart
///
/// Consulta y convierte USD/MXN usando los catálogos de Banxico.
///
/// Uso:
///   dart run example/tipo_cambio_app.dart
library;

import 'package:catalogmx/catalogmx.dart';

void printHeader() {
  print('');
  print('╔══════════════════════════════════════════════════════╗');
  print('║     💵 Tipo de Cambio USD/MXN - catalogmx            ║');
  print('╚══════════════════════════════════════════════════════╝');
  print('');
}

String formatCurrency(double amount, {int decimals = 2}) {
  return amount.toStringAsFixed(decimals);
}

void showActual() {
  print('  📊 Tipo de Cambio Actual');
  print('  ─────────────────────────');

  final actual = TipoCambioUsdCatalog.getActual();

  if (actual == null) {
    print('  ❌ No se pudo obtener el tipo de cambio');
    return;
  }

  final tc = actual['tipo_cambio'] as double;
  final fecha = actual['fecha'] as String;

  print('  📅 Fecha:        $fecha');
  print('  💱 Tipo cambio:  \$${formatCurrency(tc, decimals: 4)} MXN por 1 USD');
  print('');
  print('  Ejemplos de conversión:');
  print('    \$1 USD     = \$${formatCurrency(tc)} MXN');
  print('    \$100 USD   = \$${formatCurrency(tc * 100)} MXN');
  print('    \$1,000 USD = \$${formatCurrency(tc * 1000)} MXN');
  print('');
  print('    \$100 MXN   = \$${formatCurrency(100 / tc)} USD');
  print('    \$1,000 MXN = \$${formatCurrency(1000 / tc)} USD');
}

void showHistorico(String fecha) {
  print('  📅 Tipo de Cambio del $fecha');
  print('  ─────────────────────────────');

  final tc = TipoCambioUsdCatalog.getByFecha(fecha);

  if (tc == null) {
    print('  ❌ No hay datos para $fecha');
    print('     (Solo días hábiles tienen tipo de cambio)');
    return;
  }

  final valor = tc['tipo_cambio'] as double;
  print('  💱 Tipo cambio: \$${formatCurrency(valor, decimals: 4)} MXN/USD');
}

void convertir(double cantidad, String direccion) {
  final actual = TipoCambioUsdCatalog.getActual();

  if (actual == null) {
    print('  ❌ No se pudo obtener el tipo de cambio');
    return;
  }

  final tc = actual['tipo_cambio'] as double;

  if (direccion == 'usd_mxn') {
    final resultado = cantidad * tc;
    print('  💵 \$${formatCurrency(cantidad)} USD = \$${formatCurrency(resultado)} MXN');
  } else {
    final resultado = cantidad / tc;
    print('  🇲🇽 \$${formatCurrency(cantidad)} MXN = \$${formatCurrency(resultado)} USD');
  }
  print('     (Tipo de cambio: ${formatCurrency(tc, decimals: 4)})');
}

void showEstadisticas() {
  print('  📊 Estadísticas Recientes');
  print('  ─────────────────────────');

  final datos = TipoCambioUsdCatalog.getAll();

  if (datos.isEmpty) {
    print('  ❌ No hay datos disponibles');
    return;
  }

  // Últimos 30 registros
  final ultimos = datos.length > 30 ? datos.sublist(datos.length - 30) : datos;
  final valores = ultimos.map((d) => d['tipo_cambio'] as double).toList();

  final minimo = valores.reduce((a, b) => a < b ? a : b);
  final maximo = valores.reduce((a, b) => a > b ? a : b);
  final promedio = valores.reduce((a, b) => a + b) / valores.length;
  final primero = valores.first;
  final ultimo = valores.last;
  final variacion = ((ultimo - primero) / primero) * 100;

  print('  Registros analizados: ${ultimos.length}');
  print('  Período: ${ultimos.first['fecha']} a ${ultimos.last['fecha']}');
  print('');
  print('  Mínimo:    \$${formatCurrency(minimo, decimals: 4)}');
  print('  Máximo:    \$${formatCurrency(maximo, decimals: 4)}');
  print('  Promedio:  \$${formatCurrency(promedio, decimals: 4)}');
  print('  Variación: ${variacion >= 0 ? '+' : ''}${formatCurrency(variacion)}%');
}

void main(List<String> args) {
  printHeader();

  final comando = args.isNotEmpty ? args[0].toLowerCase() : 'actual';

  switch (comando) {
    case 'actual':
      showActual();
      break;

    case 'historico':
      final fecha = args.length > 1
          ? args[1]
          : DateTime.now().toIso8601String().split('T')[0];
      showHistorico(fecha);
      break;

    case 'usd':
      final cantidad = args.length > 1 ? double.parse(args[1]) : 100.0;
      convertir(cantidad, 'usd_mxn');
      break;

    case 'mxn':
      final cantidad = args.length > 1 ? double.parse(args[1]) : 1000.0;
      convertir(cantidad, 'mxn_usd');
      break;

    case 'stats':
      showEstadisticas();
      break;

    default:
      print('  Uso:');
      print('    dart run example/tipo_cambio_app.dart actual');
      print('    dart run example/tipo_cambio_app.dart historico 2024-06-15');
      print('    dart run example/tipo_cambio_app.dart usd 100');
      print('    dart run example/tipo_cambio_app.dart mxn 1000');
      print('    dart run example/tipo_cambio_app.dart stats');
  }

  print('');
}
